packages/examples/test/core/memoize.tsx

```diff
@@ -0,0 +1,186 @@
+import * as AI from 'ai-jsx';
+
+it('ensures that elements are only rendered once', async () => {
+  let didRender = false;
+  function Component() {
+    if (didRender) {
+      return 'FAIL';
+    }
+
+    didRender = true;
+    return 'PASS';
+  }
+
+  const ctx = AI.createRenderContext();
+  const element = ctx.memo(<Component />);
+  expect(await ctx.render(element)).toBe('PASS');
+  expect(await ctx.render(element)).toBe('PASS');
+});
+
+it('works with nested components', async () => {
+  let didRender = false;
+  function Component() {
+    if (didRender) {
+      return 'FAIL';
+    }
+
+    didRender = true;
+    return 'PASS';
+  }
+
+  function Parent() {
+    return (
+      <>
+        <Component />
+      </>
+    );
+  }
+
+  const ctx = AI.createRenderContext();
+  const element = ctx.memo(<Parent />);
+  expect(await ctx.render(element)).toBe('PASS');
+  expect(await ctx.render(element)).toBe('PASS');
+});
+
+it('works with nested/async components', async () => {
+  let didRender = false;
+  function Component() {
+    if (didRender) {
+      return 'FAIL';
+    }
+
+    didRender = true;
+    return 'PASS';
+  }
+
+  function AsyncParent() {
+    return Promise.resolve(
+      <>
+        <Component />
+      </>
+    );
+  }
+
+  const ctx = AI.createRenderContext();
+  const element = ctx.memo(<AsyncParent />);
+  expect(await ctx.render(element)).toBe('PASS');
+  expect(await ctx.render(element)).toBe('PASS');
+});
+
+it('works for streams', async () => {
+  async function* Component() {
+    yield 3;
+    yield 2;
+    yield 1;
+    return 'LIFTOFF';
+  }
+
+  const ctx = AI.createRenderContext();
+  const element = ctx.memo(<Component />);
+
+  const frames = [] as string[];
+  const renderResult = ctx.render(element);
+  for await (const frame of renderResult) {
+    frames.push(frame);
+  }
+  expect(frames).toEqual(['3', '2', '1']);
+  expect(await renderResult).toBe('LIFTOFF');
+  expect(await ctx.render(element)).toBe('LIFTOFF');
+});
+
+it('works for append-only streams', async () => {
+  async function* Component() {
+    yield AI.AppendOnlyStream;
+    yield 3;
+    yield 2;
+    yield 1;
+    return 'LIFTOFF';
+  }
+
+  const ctx = AI.createRenderContext();
+  const element = ctx.memo(<Component />);
+
+  const frames = [] as string[];
+  const renderResult = ctx.render(element);
+  for await (const frame of renderResult) {
+    frames.push(frame);
+  }
+  expect(frames).toEqual(['', '3', '32', '321']);
+  expect(await renderResult).toEqual('321LIFTOFF');
```

Comment:
lol ugh re: frames not having the final result. (No need to change it now.)

Reply:
the best